os/exec/process_monitor.go

```diff
@@ -0,0 +1,286 @@
+	}
+	return nil
+}
+
+func (m EnvMap) toSlice() []string {
+	if m == nil || len(m) == 0 {
+		return nil
+	}
+	envVars := os.Environ()
+	for k, v := range m {
+		envVars = append(envVars, fmt.Sprintf("%s=%s", k, v))
+	}
+	return envVars
+}
+
+type processListener struct {
+	complete func()
+	err      func(error)
+}
+
+// NewProcessListener returns a new ProcessListener
+func NewProcessListener(
+	complete func(),
+	err func(error),
+) ProcessListener {
+	return &processListener{
+		complete: complete,
+		err:      err,
+	}
+}
+
+func (pl *processListener) OnComplete() {
+	if fn := pl.complete; fn != nil {
+		fn()
+	}
+}
+
+func (pl *processListener) OnError(err error) {
+	if fn := pl.err; fn != nil {
+		fn(err)
+	}
+}
+
+type startFn func()
+
+type processMonitor struct {
+	sync.Mutex
+	cmd        *oexec.Cmd
+	cancelFunc context.CancelFunc
+	stdoutFd   *os.File
+	stderrFd   *os.File
+	startFn    startFn
+	listener   ProcessListener
+	err        error
+	running    bool
+	done       bool
+}
+
+// NewProcessMonitor creates a new ProcessMonitor
+func NewProcessMonitor(cmd Cmd, pl ProcessListener) (ProcessMonitor, error) {
+	if err := cmd.validate(); err != nil {
+		return nil, err
+	}
+
+	fileInfo, err := os.Stat(cmd.OutputDir)
+	if err != nil {
+		return nil, fmt.Errorf("specified directory does not exist: %v", err)
+	}
+
+	if !fileInfo.IsDir() {
+		return nil, fmt.Errorf("specified path is not a directory: %v", cmd.OutputDir)
+	}
+
+	base := filepath.Base(cmd.Path)
+	stdoutWriter, err := newWriter(cmd.OutputDir, base, defaultStdoutSuffix)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open stdout writer: %v", err)
+	}
+
+	stderrWriter, err := newWriter(cmd.OutputDir, base, defaultStderrSuffix)
+	if err != nil {
+		stdoutWriter.Close()
+		return nil, fmt.Errorf("unable to open stderr writer: %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	ocmd := oexec.CommandContext(ctx, cmd.Path)
+	ocmd.Args = cmd.Args
+	ocmd.Dir = cmd.OutputDir
+	ocmd.Stderr = stderrWriter
+	ocmd.Stdout = stdoutWriter
+	ocmd.Env = cmd.Env.toSlice()
+
+	pm := &processMonitor{
+		listener:   pl,
+		stdoutFd:   stdoutWriter,
+		stderrFd:   stderrWriter,
+		cancelFunc: cancel,
+		cmd:        ocmd,
+	}
+	pm.startFn = pm.startAsync
+
+	return pm, nil
+}
+
+func (pm *processMonitor) notifyListener(err error) {
+	if pm.listener == nil {
+		return
+	}
+	if err == nil {
+		pm.listener.OnComplete()
+	} else {
+		pm.listener.OnError(err)
+	}
+}
+
+func (pm *processMonitor) Start() error {
+	pm.Lock()
+	if pm.done {
+		pm.Unlock()
+		return errUnableToStartClosed
+	}
+
+	if pm.err != nil {
+		pm.Unlock()
+		return pm.err
+	}
+
+	if pm.running {
+		pm.Unlock()
+		return errUnableToStartRunning
+	}
+	pm.Unlock()
+
+	pm.startFn()
+	return nil
+}
+
+func (pm *processMonitor) startAsync() {
+	go pm.startSync()
+}
+
+func (pm *processMonitor) startSync() {
+	pm.Lock()
+	pm.running = true
+	pm.Unlock()
+
+	if err := pm.cmd.Run(); err != nil {
```

Comment:
can the err!= nil and err == nil case be merged? it seems they are the same regardless the result for err?